compiler/src/lnntoamm/Types.ts

```diff
@@ -85,54 +103,85 @@ export class Builtin extends Type {
     return this;
   }
 
-  compatibleWithConstraint(ty: Type): boolean {
+  compatibleWithConstraint(ty: Type, scope: Scope): boolean {
     if (ty instanceof Builtin) {
       return this.eq(ty);
-    } if (ty instanceof OneOf || ty instanceof Generated) {
-      return ty.compatibleWithConstraint(this);
+    } else if (ty instanceof OneOf || ty instanceof Generated) {
+      return ty.compatibleWithConstraint(this, scope);
+    } else if (ty instanceof HasOperator) {
+      return Has.operator(ty, scope, this).length !== 0;
+    } else if (ty instanceof HasMethod) {
+      return Has.method(ty, scope, this).length !== 0;
+    } else if (ty instanceof HasField) {
+      return false;
+    } else if (ty instanceof Interface) {
+      return ty.fields.length === 0 &&
```

Comment:
Same comment on the `fields` check here looking very temporary but not having an accompanying `TODO` comment.